### PR TITLE
Upstream some list functions.

### PR DESCRIPTION
**Pull request by Geometer1729**
This PR includes some functions for dealing with lists and property tests for those functions.

#274 and #304 were both closed to change the source branch.

### CODE REVIEW

File: Plutarch/List.hs [hunk 1]
```diff
@@ -363,3 +370,134 @@ plistEquals =
...
+          PSCons y ys' ->
+            pif
+              (x #<= y)
+              (pcons # x #$ self # xs' # ys)
+              (pcons # y #$ self # xs # ys')
+
+-- most of the tricks from timsort are not implemented here
+-- just the central strategy of finding runs of
+-- consecutive data and then merging them together
+ptimSort :: POrd a => Term s (PList a :--> PList a)
```
Comment (L-as) [inline Plutarch/List.hs:453]: Why do we need this?

File: Plutarch/List.hs [hunk 2]
```diff
@@ -363,3 +370,134 @@ plistEquals =
...
+      ( \y ys ->
+          pif
+            (f # y)
+            (pcon $ PJust y)
+            (self # f # ys)
+      )
+      (pcon PNothing)
+      xs
+
+psort :: POrd a => Term s (PList a :--> PList a)
```
Comment (L-as) [inline Plutarch/List.hs:407]: When do you actually use sorting?

File: plutarch.cabal [hunk 3]
```diff
@@ -169,12 +169,14 @@ test-suite examples
...
     , bytestring
     , generics-sop
+    , hedgehog
     , plutarch
     , plutus-core
     , plutus-ledger-api
     , plutus-tx
     , rank2classes       >=1.4.4
     , tasty
+    , tasty-hedgehog
```
Comment (L-as) [inline plutarch.cabal:179]: We've committed to using sydtest rather than tasty.

File: plutarch-test/src/Plutarch/Test/Property/Util.hs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import Control.Monad.IO.Class (liftIO)
+import Data.Proxy (Proxy (..))
+import Data.Text (Text)
+
+import Plutus.V1.Ledger.Scripts (Script (..), ScriptError)
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget)
+
+import Hedgehog (Gen, Property, PropertyT, annotate, annotateShow, assert, forAll, property, (===))
+
+class Marshal h (p :: PType) | h -> p where
```
Comment (L-as) [inline plutarch-test/src/Plutarch/Test/Property/Util.hs:37]: How is this class related to this PR?

**Review by L-as — CHANGES_REQUESTED** [review]
.

**Review by DarthPJB — COMMENTED** [review]
I can't speak to @L-as 's questions, however the Nix code changes are acceptable.

File: plutarch.cabal (hunk 3 above)
Comment (Geometer1729) [reply plutarch.cabal]: sorry about that, I forgot to remove the import when moving to sydtest.

File: plutarch-test/src/Plutarch/Test/Property/Util.hs (hunk 4 above)
Comment (Geometer1729) [reply plutarch-test/src/Plutarch/Test/Property/Util.hs:37]: When testing agreement between Haskell functions and Plutarch functions it's convenient to be able to generate the arguments to the function as Haskell terms and marshal them into Plutarch terms, that way you have both representations and can run both functions. Marshal is also used to turn the output of the Haskell function into Plutarch to test equality at the end.

File: Plutarch/List.hs (hunk 2 above)
Comment (Geometer1729) [reply Plutarch/List.hs:407]: I originally implemented this because I thought we would need sorting to convert values into sorted list so that functions like unionWith can be more efficient. Hopefully the values come sorted or can be required to be sorted, but we haven't tested that yet.

File: Plutarch/List.hs (hunk 1 above)
Comment (Geometer1729) [reply Plutarch/List.hs:453]: I was worried that merge sort might be unnecessarily slow if a lot of the lists it gets called with are nearly sorted. I plan to benchmark the project using both and see what the tradeoff looks like.

**Comment by L-as** [comment]
Two thoughts:
- I don't think we should have sorting honestly, because all parts of ScriptContext are either pre-sorted, or can be required to be sorted rather than actually sorting them. We discussed this in an internal meeting some time ago. We should make plutarch-extra with the helpers needed for this.
- I'll take a deeper look at Marshal

**Review by nixinator — COMMENTED** [review]
don't eat hedge hogs , they are engaged, no matter how tasty they are.

**Comment by TotallyNotChase** [comment]
I stumbled upon the lack of efficient sorting options when implementing `Semigroup` for `PValue`. That could use `unionWith`, and `unionWith` could benefit from sorted lookups and such. I can't speak to whether or not that belongs in this repo but `Semigroup (Term s PValue)` (or simply `unionWith`) is sort of high priority IMO.

**Comment by Geometer1729** [comment]
> I stumbled upon the lack of efficient sorting options when implementing `Semigroup` for `PValue`. That could use `unionWith`, and `unionWith` could benefit from sorted lookups and such. I can't speak to whether or not that belongs in this repo but `Semigroup (Term s PValue)` (or simply `unionWith`) is sort of high priority IMO.

This was for the same reason (in fact we even have a monoid instance for PValue we want to upstream) but if Las is right and the values can be required to come sorted then unionWith can pretty easily preserver sortedness.

**Comment by TotallyNotChase** [comment]
> > I stumbled upon the lack of efficient sorting options when implementing `Semigroup` for `PValue`. That could use `unionWith`, and `unionWith` could benefit from sorted lookups and such. I can't speak to whether or not that belongs in this repo but `Semigroup (Term s PValue)` (or simply `unionWith`) is sort of high priority IMO.
> 
> This was for the same reason (in fact we even have a monoid instance for PValue we want to upstream) but if Las is right and the values can be required to come sorted then unionWith can pretty easily preserver sortedness.

Do you think it's possible to upstream some of these `Value` utilities (not necessarily instances, just `unionWith` and similar) to another library that people can target? I actually need this stuff as well and I've been handwriting all of this. There's no reason for all users to go through this much definitions just to be able to use maps..... of all things. We could have a place for utility functions now that we can merge to `plutarch-extra` later as Las suggests.

**Comment by emiflake** [comment]
+1 on having a clear place to upstream utility functions pre-`plutarch-extra`. What's blocking getting `plutarch-extra` going, though?

**Comment by TotallyNotChase** [comment]
screw it, let's just get plutarch-extra ready alongside 1.2 release

@MangoIV could you make the initial plutarch-extra scaffold in this repo please? It should be a separate library - doesn't necessarily need to contain any real code immediately.

@Geometer1729 could you prepare the PRs for the code of this please? Most of the functions in this PR should also go to `plutarch-extra`, and whatever else utilities necessary - be sure to separate the PRs though. However, you should add the instances and utilities for `PMap` (and `PValue`) in Plutarch itself. We'll then have the fundamental utilities for the fundamental data types - though we're probably missing a few for `Plutarch.Pair`.

@L-as are you ok with this plan? anything you want to change?

Relevant issues
* https://github.com/Plutonomicon/plutarch/issues/139
* https://github.com/Plutonomicon/plutarch/issues/190

please make sure to mark any relevant PRs with this issues.

**Comment by MangoIV** [comment]
Sure, I can do that, I'm currently working on #303 but I can do this at <some idea of a deadline you insert ;)>.

**Comment by L-as** [comment]
@TotallyNotChase sounds good to me

**Comment by L-as** [comment]
#357

**Comment by L-as** [comment]
Closing this in favor of #357
